### PR TITLE
Correct send_to_components

### PR DESCRIPTION
**Pull request by peterbarker**
Attempts to solve a problem where component messages aren't send with the correct wire protocol.

Tested only as far as "it compiles".

Hopefully closes https://github.com/ArduPilot/ardupilot/issues/12567

This may cause vehicles to explode and/or turn into small mounds of granulated gelatin.  It's unlikely, however.

### CODE REVIEW

File: libraries/GCS_MAVLink/MAVLink_routing.h [hunk 1]
```diff
@@ -63,4 +63,6 @@ class MAVLink_routing
 
     // special handling for heartbeat messages
     void handle_heartbeat(mavlink_channel_t in_channel, const mavlink_message_t &msg);
+
+    void send_to_components(const char *pkt, const mavlink_msg_entry_t *entry);
```
Comment (rmackay9) [inline libraries/GCS_MAVLink/MAVLink_routing.h]: should we add a comment above this function?  the arguments are a little harder to understand than the previous one.  At least I'm not familiar with what a mavlink_msg_entry_t is.

**Comment by rmackay9** [comment]
I've tested with the Drotag and found that sadly this doesn't make the [Drotag geotagger](http://ardupilot.org/copter/docs/common-geotagging-drotagx.html) work.  That doesn't mean that there is anything wrong with this PR however, it may still be some other issue within the Drotag.

File: libraries/GCS_MAVLink/MAVLink_routing.h (hunk 1 above)
Comment (peterbarker) [reply libraries/GCS_MAVLink/MAVLink_routing.h]: Commend added.

**Comment by pklapperich** [comment]
@peterbarker I'm not sure if you saw, but I [left a comment](https://github.com/ArduPilot/ardupilot/issues/12567#issuecomment-548178348) over on #12567 after testing this. It definitely fixed the problem I had. It also seems to add automatic mavlink version detection (my mavlink1 only camera works whether I configure the serial port in ArduCopter to Mavlink1 or Mavlink2). If that's intentional, then I think it's good.

**Comment by peterbarker** [comment]
@pklapperich thanks for the testing report, glad it worked :-)

Marked as devcall to see if we can merge it today.

File: libraries/GCS_MAVLink/MAVLink_routing.h [hunk 2]
```diff
@@ -32,8 +32,11 @@ class MAVLink_routing
     /*
       send a MAVLink message to all components with this vehicle's system id
       This is a no-op if no routes to components have been learned
+
+      msgid here is the mavlink message ID, pkt is a pointer to a
+      mavlink message structure (e.g. a mavlink_command_long_t)
     */
-    void send_to_components(const mavlink_message_t &msg);
+    void send_to_components(uint32_t msgid, const char *pkt);
```
Comment (tridge) [inline libraries/GCS_MAVLink/MAVLink_routing.h]: I'd like a length arg, and chec

Comment (peterbarker) [reply libraries/GCS_MAVLink/MAVLink_routing.h]: I've added a an assert in SITL and we pass through the smaller of the passed packet length and the maximum message length.

**Comment by rmackay9** [comment]
Another one for @tridge to confirm he is happy with before we merge

**Comment by rmackay9** [comment]
this has been added to Copter-4.0.0-rc3

**Pull request merged**